### PR TITLE
bot filtering

### PR DESCRIPTION
**Pull request by msohailhussain**

### CODE REVIEW

**Comment by msohailhussain** [comment]
@tylerbrandt IN the Fullstack compatability, it is checking out branch with commit id but it's not working, can u plz verify?

**Comment by msohailhussain** [comment]
build

**Comment by msohailhussain** [comment]
build

File: OptimizelySDK.Tests/EventTests/EventBuilderTest.cs [hunk 1]
```diff
@@ -1,3 +1,19 @@
+/* 
+ * Copyright 2017-2018, Optimizely
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/EventBuilderTest.cs]: Thanks for adding this.

File: OptimizelySDK.Tests/EventTests/EventBuilderTest.cs [hunk 2]
```diff
@@ -16,7 +32,7 @@ public class EventBuilderTest
         private string TestUserId = string.Empty;
         private ProjectConfig Config;
         private EventBuilder EventBuilder;
-
+
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/EventBuilderTest.cs]: nit. Seems unnecessary spaces got put in there. Lets undo.

File: OptimizelySDK/Logger/ILogger.cs [hunk 3]
```diff
@@ -20,6 +20,7 @@ public enum LogLevel
         DEBUG,
         INFO,
         ERROR,
+        WARN,
```
Comment (aliabbasrizvi) [inline OptimizelySDK/Logger/ILogger.cs]: WARN should go above ERROR IMO

Comment (aliabbasrizvi) [reply OptimizelySDK/Logger/ILogger.cs]: Also, update year

File: OptimizelySDK/ProjectConfig.cs [hunk 4]
```diff
@@ -552,5 +559,32 @@ public Rollout GetRolloutFromId(string rolloutId)
             ErrorHandler.HandleError(new Exceptions.InvalidRolloutException("Provided rollout is not in datafile."));
             return new Rollout();
         }
+
+        /// <summary>
+        /// Get attribute ID for the provided attribute key
+        /// </summary>
+        /// <param name="attributeKey">Key of the Attribute</param>
+        /// <returns>Attribute ID corresponding to the provided attribute key. Attribute key if it is a reserved attribute</returns>
+        public string GetAttributeId(string attributeKey)
```
Comment (aliabbasrizvi) [inline OptimizelySDK/ProjectConfig.cs]: Please add unit tests for this new module

File: OptimizelySDK.Tests/TestData.json [hunk 5]
```diff
@@ -634,5 +634,6 @@
 		"key": "purchase"
 	}],
 	"revision": "15",
-	"anonymizeIP": "false"
+	"anonymizeIP": "false",
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/TestData.json]: This will always be a boolean value

File: OptimizelySDK.Tests/TestData.json [hunk 6]
```diff
@@ -634,5 +634,6 @@
...
+	"botFiltering": "true"
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/TestData.json]: This will always be a boolean value.

File: OptimizelySDK.Tests/EventTests/EventBuilderTest.cs [hunk 7]
```diff
@@ -894,5 +995,371 @@ public void TestCreateImpressionEventWithBucketingIDAttribute()
...
+                            },
+                            { "visitor_id", TestUserId }
+                        }
+                    }
+                },
+                {"project_id", "7720880029" },
+                {"account_id", "1592310167" },
+                {"client_name", "csharp-sdk" },
+                {"client_version", Optimizely.SDK_VERSION },
+                {"revision", 15 },
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/EventBuilderTest.cs]: Can you point me to where revision is set? It should be what it is in the datafile i.e. it should be string.

Comment (msohailhussain) [reply OptimizelySDK.Tests/EventTests/EventBuilderTest.cs]: https://github.com/optimizely/csharp-sdk/blob/master/OptimizelySDK/ProjectConfig.cs#L31
https://github.com/optimizely/csharp-sdk/blob/master/OptimizelySDK/ProjectConfig.cs#L266
We are deserializing it and it is set automatically.

Comment (aliabbasrizvi) [reply OptimizelySDK.Tests/EventTests/EventBuilderTest.cs]: Bot revision and version need to be string and not int. Can you please fix that as well.

Thanks.

File: OptimizelySDK/ProjectConfig.cs (hunk 4 above)
Comment (mfahadahmed) [reply OptimizelySDK/ProjectConfig.cs]: Already covered here:

File: OptimizelySDK/Event/Builder/EventBuilder.cs [hunk 8]
```diff
@@ -35,7 +35,9 @@ public class EventBuilder
 
         private const string ACTIVATE_EVENT_KEY = "campaign_activated";
 
-        public const string RESERVED_ATTRIBUTE_KEY_BUCKETING_ID_EVENT_PARAM_KEY = "optimizely_bucketing_id";
+        public const string BOT_FILTERING_ATTRIBUTE = "$opt_bot_filtering";
+
+        public const string USER_AGENT_ATTRIBUTE = "$opt_user_agent";
```
Comment (aliabbasrizvi) [inline OptimizelySDK/Event/Builder/EventBuilder.cs]: nit. Like in other SDKs, lets consolidate all of the "special" attributes i.e. bucketing ID, bot filtering and user agent attributes.

File: OptimizelySDK/ProjectConfig.cs (hunk 4 above)
Comment (aliabbasrizvi) [reply OptimizelySDK/ProjectConfig.cs]: Oh, good. GitHub does not render large files by default and hence missed it.

File: OptimizelySDK.Tests/ProjectConfigTest.cs [hunk 9]
```diff
@@ -810,5 +812,55 @@ public void TestVariationFeatureEnabledProperty()
...
+        }
+
+        [Test]
+        public void TestGetAttributeIdWithReservedPrefix()
+        {
+            // Verify that attribute key is returned for reserved attribute key.
+            Assert.AreEqual(Config.GetAttributeId(EventBuilder.USER_AGENT_ATTRIBUTE), EventBuilder.USER_AGENT_ATTRIBUTE);
+
+            // Verify that attribute Id is returned for attribute key with reserved prefix that does not exist in datafile.
+            Assert.AreEqual(Config.GetAttributeId(EventBuilder.USER_AGENT_ATTRIBUTE), EventBuilder.USER_AGENT_ATTRIBUTE);
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/ProjectConfigTest.cs]: nit. This test is no different from the previous test. Let's have a separate example. Like lets test for $opt_reserved_prefix_attribute

File: OptimizelySDK.Tests/TestData.json [hunk 10]
```diff
@@ -634,5 +634,6 @@
 		"key": "purchase"
 	}],
 	"revision": "15",
-	"anonymizeIP": "false"
+	"anonymizeIP": false,
+	"botFiltering": true
 }
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/TestData.json]: nit. Add new line at end of file.

File: OptimizelySDK/ProjectConfig.cs [hunk 11]
```diff
@@ -552,5 +559,32 @@ public Rollout GetRolloutFromId(string rolloutId)
...
+                    Logger.Log(LogLevel.WARN, $@"Attribute {attributeKey} unexpectedly has reserved prefix {RESERVED_ATTRIBUTE_PREFIX}; using attribute ID instead of reserved attribute name.");
+
+                return attribute.Id;
+            }
+            else if (hasReservedPrefix)
+            {
+                return attributeKey;
+            }
+
+            Logger.Log(LogLevel.ERROR, $@"Attribute key ""{attributeKey}"" is not in datafile.");
```
Comment (aliabbasrizvi) [inline OptimizelySDK/ProjectConfig.cs]: nit. Just single pair of double quotes should be good right?

File: OptimizelySDK/Utils/ReservedAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace OptimizelySDK.Utils
+{
+    public class ReservedAttribute
```
Comment (aliabbasrizvi) [inline OptimizelySDK/Utils/ReservedAttribute.cs]: nit. Call this `ControlAttribute` like in other SDKs

File: OptimizelySDK/Utils/ReservedAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public const string BOT_FILTERING_ATTRIBUTE = "$opt_bot_filtering";
```
Comment (aliabbasrizvi) [inline OptimizelySDK/Utils/ReservedAttribute.cs]: nit. Alphabetize.

**Pull request merged**